fix(tester.py): adjust reliable_replication_factor for tablets

yarongilor opened the pull request:
In case tablets are enabled, it's better to set RF smaller than dc-nodes-number,
	so decommission is allowed.
	fixes: #7620

### Testing
<!-- Add links to Argus/Jenkins of test test done with this PR -->
<!-- This would help the reviewer to cross check what was tested, and and review the results as needed -->
- [ ]

### PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [ ] I added the relevant `backport` labels
- [ ] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

yarongilor commented:
Not tested yet.

yarongilor commented:
@roydahan , an alternative solution could be modifying nodetool-decommission nemesis to add node before decommission. But it might have more impact and risk for unexpected results.

yarongilor commented:
The nodetool decommission run successfully, but the nemesis failed adding a new node for:
```
2024-06-10 14:42:55.517 <2024-06-10 14:42:55.290>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=5984a639-583b-4c7c-a722-340899a7099f during_nemesis=NodetoolDecommission: type=DATABASE_ERROR regex=(^ERROR|!\s*?ERR).*\[shard.*\] line_number=2006 node=longevity-large-partitions-200k-pks-db-node-d35607de-0-6
2024-06-10T14:42:55.290+00:00 longevity-large-partitions-200k-pks-db-node-d35607de-0-6      !ERR | scylla[7673]:  [shard  0:main] init - Startup failed: seastar::rpc::remote_verb_error (group [780f3231-2730-11ef-928d-083760ea7146] raft operation [add_entry] timed out)
```
@aleksbykov , is there an SCT fix for this raft limitation error?

yarongilor commented:
(Addresses https://github.com/scylladb/scylla-cluster-tests/issues/7462 as well)

roydahan commented:
The question is if it's sufficient, we use this method in `_prepare_test_table` which is being used only in some specific places.
I'm sure there are other tests configurations were we have 3 nodes and RF=3.

I suggest that the nemesis will be treated as "best effort" and on top of this change if the decommission can't run due to this limitation is should be skipped without producing an error.

Regardless, this change should be tested extensively in several test scenarios.

yarongilor commented:
> The question is if it's sufficient, we use this method in `_prepare_test_table` which is being used only in some specific places. I'm sure there are other tests configurations were we have 3 nodes and RF=3.
> 
> I suggest that the nemesis will be treated as "best effort" and on top of this change if the decommission can't run due to this limitation is should be skipped without producing an error.
> 
> Regardless, this change should be tested extensively in several test scenarios.

ok, so in addition to `_prepare_test_table`, the nodetool decommission nemesis will check if an RF equal to number-of-nodes is found, then skip the nemesis.

yarongilor commented:
> > The question is if it's sufficient, we use this method in `_prepare_test_table` which is being used only in some specific places. I'm sure there are other tests configurations were we have 3 nodes and RF=3.
> > I suggest that the nemesis will be treated as "best effort" and on top of this change if the decommission can't run due to this limitation is should be skipped without producing an error.
> > Regardless, this change should be tested extensively in several test scenarios.
> 
> ok, so in addition to `_prepare_test_table`, the nodetool decommission nemesis will check if an RF equal to number-of-nodes is found, then skip the nemesis.

@roydahan , added a separate commit, please review:
```
fix(decommission nemesis): adjust for tablets
        In case tablets are enabled, skip decommission if a keyspace
        replicatin factor equals number-of-nodes.
        Added is_system_keyspace() as well.
        refs: https://github.com/scylladb/scylla-cluster-tests/issues/7620
```
i'll run some tests:
1. for https://github.com/scylladb/scylla-cluster-tests/issues/7462
2. for https://github.com/scylladb/scylla-cluster-tests/issues/7620
3. for longevity topology changes

yarongilor commented:
Running tests of:
[7620](https://argus.scylladb.com/test/3d7ecc1b-e7c3-4fb8-9178-0f397b3d9c7c/runs?additionalRuns[]=12e69b08-7eee-4c45-9fbc-67ec9fc6d45d)
[7462](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/6/)
[gemini 1T 3-nodes](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/7/)

vponomaryov commented on sdcm/utils/database_query_utils.py:
```diff
@@ -181,6 +181,45 @@ def get_table_clustering_order(ks_cf: str, ck_name: str, session) -> str:
+            if replication_factor > max_replication_factor:
+                max_replication_factor = replication_factor
+                keyspace_with_max_rf = keyspace_name
+        elif 'NetworkTopologyStrategy' in replication['class']:
```
Better to add `else` clause with some fall-back behaviour (or at least warning msg) for case these first 2 conditions are not matched.

yarongilor replied:
fixed

vponomaryov commented on sdcm/utils/database_query_utils.py:
```diff
@@ -181,6 +181,45 @@ def get_table_clustering_order(ks_cf: str, ck_name: str, session) -> str:
+    cql_result = session.execute(query)
+
+    max_replication_factor = 0
+    keyspace_with_max_rf = None
```
`keyspace_with_max_rf` as `None` is not expected in above code logic.
Probably set it to some string?

yarongilor replied:
fixed

fruch commented:
> Running tests of:
> [7620](https://argus.scylladb.com/test/3d7ecc1b-e7c3-4fb8-9178-0f397b3d9c7c/runs?additionalRuns[]=12e69b08-7eee-4c45-9fbc-67ec9fc6d45d) , [7620 test2](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/longevity-large-partition-200k-pks-4days-gce-test/7/)
> [7462](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/6/)
> [gemini 1T 3-nodes](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/7/)

You mentioned tests here, a month ago

But they reported back if they were successful or not.

It's kind of hard to understand what is the situation in this one, it's not enough to put someone on review, some communication is needed for things to move along

yarongilor commented on sdcm/tester.py:
```diff
@@ -753,7 +754,10 @@ def reliable_replication_factor(self) -> int:
+        min_nodes_dc = min([int(nodes_num) for nodes_num in n_db_nodes.split() if int(nodes_num) > 0])
+        with self.db_cluster.cql_connection_patient(self.db_cluster.nodes[0]) as session:
+            # In case tablets are enabled, it's better to set RF smaller than dc-nodes-number, so decommission is allowed.
+            return max([min_nodes_dc - 1, 1]) if is_tablets_feature_enabled(session) else min_nodes_dc
```
there are 105 yaml with 3 nodes:
```
$ grep -Eri 'n_db_nodes: 3' --include \*.yaml . | wc -l
105
```
`RF=2` is exactly what needed in this case, in order to allow a decommission.

fruch replied:
@yarongilor 

do we still need this ? if we know how to drop the RF in runtime ?

yarongilor commented:
> > Running tests of:
> > [7620](https://argus.scylladb.com/test/3d7ecc1b-e7c3-4fb8-9178-0f397b3d9c7c/runs?additionalRuns%5B%5D=12e69b08-7eee-4c45-9fbc-67ec9fc6d45d) , [7620 test2](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/longevity-large-partition-200k-pks-4days-gce-test/7/)
> > [7462](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/6/)
> > [gemini 1T 3-nodes](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/7/)
> 
> You mentioned tests here, a month ago
> 
> But they reported back if they were successful or not.
> 
> It's kind of hard to understand what is the situation in this one, it's not enough to put someone on review, some communication is needed for things to move along

right, these tests just failed for unrelated blocking issues. have to rerun it with master latest.

bhalevy commented on sdcm/nemesis.py:
```diff
@@ -1303,6 +1304,15 @@ def _terminate_cluster_node(self, node):
+        with self.cluster.cql_connection_patient(self.target_node) as session:
+            if is_tablets_feature_enabled(session):
+                max_rf, keyspace = get_max_replication_factor(session)
+                if max_rf == len(self.cluster.nodes):
```
Ultimately, need to check rf vs. number of nodes in the datacenter.

bhalevy replied:
With tablet at the moment you even need to check that the test doesn't try to decommission the last node in a particular rack.

yarongilor replied:
that is theoretically right. currently SCT doesn't cover such a relevant scenario.
I think it would be better to address this in a separate issue+PR.
We may also have to address the issue for AddRemoveDC nemesis.

yarongilor commented:
> > > Running tests of:
> > > [7620](https://argus.scylladb.com/test/3d7ecc1b-e7c3-4fb8-9178-0f397b3d9c7c/runs?additionalRuns%5B%5D=12e69b08-7eee-4c45-9fbc-67ec9fc6d45d) , [7620 test2](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/longevity-large-partition-200k-pks-4days-gce-test/7/)
> > > [7462](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/6/)
> > > [gemini 1T 3-nodes](https://jenkins.scylladb.com/job/scylla-staging/job/yarongilor/job/reproductions/7/)
> > 
> > 
> > You mentioned tests here, a month ago
> > But they reported back if they were successful or not.
> > It's kind of hard to understand what is the situation in this one, it's not enough to put someone on review, some communication is needed for things to move along
> 
> right, these tests just failed for unrelated blocking issues. have to rerun it with master latest.

nodetool-decommission nemesis [passed](https://argus.scylladb.com/test/3d7ecc1b-e7c3-4fb8-9178-0f397b3d9c7c/runs?additionalRuns%5B%5D=12e69b08-7eee-4c45-9fbc-67ec9fc6d45d) ok:
```
< t:2024-07-25 18:04:00,444 f:nemesis.py      l:1887 c:sdcm.nemesis         p:INFO  > sdcm.nemesis.SisyphusMonkey: >>>>>>>>>>>>>Started random_disrupt_method nodetool_decommission
< t:2024-07-25 18:06:11,305 f:remote_base.py  l:521  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.56>: Running command "echo "/usr/bin/nodetool  decommission " > /tmp/remoter_cmd_d9352e01-328f-4c85-b25a-7221e8963171.sh"...
< t:2024-07-25 18:06:11,307 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:INFO  > 2024-07-25 18:06:11.305: (NodetoolEvent Severity.NORMAL) period_type=begin event_id=31a0474c-5cbf-435e-b4d5-65afffa97696: nodetool_command=decommission node=longevity-large-partitions-200k-pks-db-node-8b9a36aa-0-5
< t:2024-07-25 18:06:11,846 f:base.py         l:143  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.56>: Command "echo "/usr/bin/nodetool  decommission " > /tmp/remoter_cmd_d9352e01-328f-4c85-b25a-7221e8963171.sh" finished with status 0
< t:2024-07-25 18:13:03,956 f:cluster.py      l:2632 c:sdcm.cluster_gce     p:DEBUG > Node longevity-large-partitions-200k-pks-db-node-8b9a36aa-0-5 [34.138.157.82 | 10.142.0.56]: Command '/usr/bin/nodetool  decommission ' duration -> 411.92979814499995 s
< t:2024-07-25 18:13:03,962 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:INFO  > 2024-07-25 18:13:03.958: (NodetoolEvent Severity.NORMAL) period_type=end event_id=31a0474c-5cbf-435e-b4d5-65afffa97696 duration=6m52s: nodetool_command=decommission node=longevity-large-partitions-200k-pks-db-node-8b9a36aa-0-5
```

yarongilor commented:
i think this could be merged.

fruch commented:
> A relevant Scylla update for this fix is that [scylladb/scylladb#16129](https://github.com/scylladb/scylladb/issues/16129) is merged. That means decreasing a keyspace RF is supported (see https://opensource.docs.scylladb.com/stable/cql/ddl.html#alter-keyspace-with-tablets-experimental). Currently, this fix causes tests with rf=3 and 3 nodes to never run a nodetool-decommission nemesis. So a possible enhancement suggestion for this PR could be: for a requested decommission do the following:
> 
> * If a keyspace with RF = num-of-nodes exist (e.g. rf = num-of-noeds = 3):
> * * alter keyspace RF to RF-1 (e.g. rf=2)
> * * decommission the requested node.
> * * finally ( or at a relevant step of nemesis ) do:
> * * * Add a new node
> * * * alter keyspace RF back to original RF value (e.g. rf=3)
>       @fruch , @roydahan , please confirm/reject/advise for this suggestion.

Sound like a better approach, that would cover this new use case.
but keep in mind, this fix might not get backported to 6.0/2024.2, and it would still suffer for those thing (maybe not that problematic)

> 
> i'll also write a new Dtest for such a scenario, since there is no any (@bhalevy please correct if it's wrong), and that's the customer best practice to decommission in such a configuration. [scylladb/scylla-dtest#4767](https://github.com/scylladb/scylla-dtest/pull/4767)

kbr-scylla commented:
Guys, this PR is opened since Jun 10, it blocks us from getting a passing longevity on https://github.com/scylladb/scylladb/issues/19031#issuecomment-2237345729, which blocks us from closing it, which blocks https://github.com/scylladb/scylladb/issues/19105 from getting closed, which is a P1 issue (but that's the only thing blocking it), and @kostja is eating my brain about it

please can you increase the priority of reviewing and merging this?

tchaikov reviewed (approved):
lgtm

tchaikov commented:
@fruch could you take another look?

yarongilor commented:
> Guys, this PR is opened since Jun 10, it blocks us from getting a passing longevity on [scylladb/scylladb#19031 (comment)](https://github.com/scylladb/scylladb/issues/19031#issuecomment-2237345729), which blocks us from closing it, which blocks [scylladb/scylladb#19105](https://github.com/scylladb/scylladb/issues/19105) from getting closed, which is a P1 issue (but that's the only thing blocking it), and @kostja is eating my brain about it
> 
> please can you increase the priority of reviewing and merging this?

@kbr-scylla , @fruch , @tchaikov ,
There could be 2 options addressing this PR.

1. Merge it as it is currently (without the above suggested fix of alter-keyspace-rf)
1.1 So it is already reviewed, tested and passed.
1.2 The drawback is that some nemesis in some longevities like nodetool-decommission will always be skipped. 
2. Add the above suggested fix of alter-keyspace-rf
2.1 This will fully solve the decommission limitation so no nemesis is skipped.
2.2 The drawback is it will take more time to write, review and test it.

Please comment any preference.
Cc: @bhalevy , @pehala

kbr-scylla commented:
@yarongilor for me, I just want to recreate the conditions from your report here https://github.com/scylladb/scylladb/issues/19031#issue-2329634055

Does option 1 do that?

yarongilor commented:
> @yarongilor for me, I just want to recreate the conditions from your report here [scylladb/scylladb#19031 (comment)](https://github.com/scylladb/scylladb/issues/19031#issue-2329634055)
> 
> Does option 1 do that?

@kbr-scylla , please see comment- https://github.com/scylladb/scylladb/issues/19031#issuecomment-2288944111

kbr-scylla commented:
Thanks @yarongilor.

So I'm for option 2.

bhalevy commented:
ping @yarongilor, is this PR ready to be merged?

yarongilor commented:
> ping @yarongilor, is this PR ready to be merged?

@bhalevy , since got votes for the 2nd alternative suggested above, the PR is now revised.
It will not simply skip nemesis, but rather try decreasing RF and run decommission.
i'll send this fix in a separate commit soon.

yarongilor commented:
> > ping @yarongilor, is this PR ready to be merged?
> 
> @bhalevy , since got votes for the 2nd alternative suggested above, the PR is now revised. It will not simply skip nemesis, but rather try decreasing RF and run decommission. i'll send this fix in a separate commit soon.

Added a new commit with the relevant modifications: `fix(decommission nemesis): adjust for tablets by decreasing replication factor`

yarongilor commented:
> > > ping @yarongilor, is this PR ready to be merged?
> > 
> > 
> > @bhalevy , since got votes for the 2nd alternative suggested above, the PR is now revised. It will not simply skip nemesis, but rather try decreasing RF and run decommission. i'll send this fix in a separate commit soon.
> 
> Added a new commit with the relevant modifications: `fix(decommission nemesis): adjust for tablets by decreasing replication factor`

Added `class DataCenterTopologyRfChange`.

A local test passed ok like:
```
< t:2024-08-20 18:18:44,063 f:sct_config.py   l:2032 c:sdcm.sct_config      p:INFO  > nemesis_class_name: DecommissionMonkey
< t:2024-08-20 18:22:38,832 f:tester.py       l:1172 c:LongevityTest        p:DEBUG > Nemesis threads [{'nemesis': <class 'sdcm.nemesis.DecommissionMonkey'>, 'nemesis_selector': []}]
< t:2024-08-20 18:22:43,664 f:nemesis.py      l:416  c:sdcm.nemesis         p:INFO  > sdcm.nemesis.DecommissionMonkey: Interval: 60 s
< t:2024-08-20 18:22:43,665 f:nemesis.py      l:1141 c:sdcm.nemesis         p:DEBUG > sdcm.nemesis.DecommissionMonkey: Set current_disruption -> NodetoolDecommission
< t:2024-08-20 18:22:43,665 f:nemesis.py      l:408  c:sdcm.nemesis         p:INFO  > sdcm.nemesis.DecommissionMonkey: Current Target: Node yarongilor-db-node-67fdbc32-2 [172.17.0.4 | 172.17.0.4] with running nemesis: NodetoolDecommission
< t:2024-08-20 18:22:51,798 f:nemesis.py      l:5207 c:sdcm.nemesis         p:DEBUG > sdcm.nemesis.DecommissionMonkey: Start disruption at `2024-08-20 18:22:51.798803`
< t:2024-08-20 18:22:51,830 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:INFO  > 2024-08-20 18:22:51.828: (DisruptionEvent Severity.NORMAL) period_type=begin event_id=32b60ee8-b5df-489b-af9c-6bb9a8596b93: nemesis_name=NodetoolDecommission target_node=Node yarongilor-db-node-67fdbc32-2 [172.17.0.4 | 172.17.0.4]
< t:2024-08-20 18:22:53,134 f:common.py       l:1312 c:utils                p:DEBUG > Executing CQL 'SELECT keyspace_name, replication FROM system_schema.keyspaces' ...
< t:2024-08-20 18:22:53,134 f:common.py       l:1320 c:utils                p:DEBUG > Executing CQL 'SELECT keyspace_name, replication FROM system_schema.keyspaces' ...
< t:2024-08-20 18:22:53,136 f:replication_strategy_utils.py l:184  c:sdcm.utils.replication_strategy_utils p:DEBUG > Altering ks_3b RF with: ALTER KEYSPACE ks_3b WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':2 }
< t:2024-08-20 18:22:53,136 f:common.py       l:1312 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE ks_3b WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':2 }' ...
< t:2024-08-20 18:22:53,136 f:common.py       l:1320 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE ks_3b WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':2 }' ...
< t:2024-08-20 18:22:53,804 f:replication_strategy_utils.py l:184  c:sdcm.utils.replication_strategy_utils p:DEBUG > Altering ks_3 RF with: ALTER KEYSPACE ks_3 WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':2 }
< t:2024-08-20 18:22:53,804 f:common.py       l:1312 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE ks_3 WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':2 }' ...
< t:2024-08-20 18:22:53,804 f:common.py       l:1320 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE ks_3 WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':2 }' ...
< t:2024-08-20 18:22:55,857 f:replication_strategy_utils.py l:192  c:sdcm.utils.replication_strategy_utils p:DEBUG > Reverting keyspaces replication factor to original value of datacenter1..
< t:2024-08-20 18:22:55,925 f:replication_strategy_utils.py l:184  c:sdcm.utils.replication_strategy_utils p:DEBUG > Altering ks_3b RF with: ALTER KEYSPACE ks_3b WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':3 }
< t:2024-08-20 18:22:55,925 f:common.py       l:1312 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE ks_3b WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':3 }' ...
< t:2024-08-20 18:22:55,925 f:common.py       l:1320 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE ks_3b WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':3 }' ...
< t:2024-08-20 18:22:57,432 f:replication_strategy_utils.py l:184  c:sdcm.utils.replication_strategy_utils p:DEBUG > Altering ks_3 RF with: ALTER KEYSPACE ks_3 WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':3 }
< t:2024-08-20 18:22:57,433 f:common.py       l:1312 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE ks_3 WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':3 }' ...
< t:2024-08-20 18:22:57,433 f:common.py       l:1320 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE ks_3 WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'datacenter1':3 }' ...
< t:2024-08-20 18:22:58,384 f:nemesis.py      l:5268 c:sdcm.nemesis         p:DEBUG > sdcm.nemesis.DecommissionMonkey: NodetoolDecommission duration -> 6 s
```

yarongilor commented:
@fruch , @roydahan , @soyacz , can you please review the last 2 commits of this PR?
it already encountered 2 new severe issues:
https://github.com/scylladb/scylladb/issues/20282 , https://github.com/scylladb/scylladb/issues/20240.

fruch commented:
> @fruch , @roydahan , @soyacz , can you please review the last 2 commits of this PR? it already encountered 2 new severe issues: [scylladb/scylladb#20282](https://github.com/scylladb/scylladb/issues/20282) , [scylladb/scylladb#20240](https://github.com/scylladb/scylladb/issues/20240).
> 
> A [test for decommission-related nemeses](https://argus.scylladb.com/test/e6972842-591d-472f-9e39-f196d3670053/runs?additionalRuns%5B%5D=4f0cedfd-0e8a-4bfd-9de7-115a32ff93da) passed ok, testing a single DC configuration. It had 3 nodes and RF=3 for stress keyspaces, so on each nemesis, the RF value was altered to 2, then back to 3 after a new node is added.
> 
> Testing a decommission in multi-DC seems to be blocked by the 2 above issues. Commits details:
> 
> * In case tablets are enabled, decommission is not permitted if a keyspace
>   replicatin factor equals number-of-nodes.
>   In this case the keyspace replication-factor should be decreased first.
> * When replication-factor is increased after a new node was added,
>   wait for tablets migration to complete.
> 
> cc: @bhalevy

@yarongilor just to clarify, those two issues doesn't happens without those changes ?

fruch commented on sdcm/utils/replication_strategy_utils.py:
```diff
@@ -188,12 +190,15 @@ def _alter_keyspace_rf(self, keyspace: str, replication_factor: int, session):
-            for keyspace in self.decreased_rf_keyspaces:
-                self._alter_keyspace_rf(keyspace=keyspace, replication_factor=self.original_nodes_number,
-                                        session=session)
+    def revert_to_original_keyspaces_rf(self, wait_node_balance: 'BaseNode' = None):
```
`wait_node_balance` sounds like a boolean, it's a bit surprising to see it's type is `BaseNode`

yarongilor replied:
fixed: `node_to_wait_for_balance`

fruch commented on sdcm/utils/replication_strategy_utils.py:
```diff
@@ -120,3 +121,98 @@ def __call__(self, **keyspaces: ReplicationStrategy) -> None:
+
+class DataCenterTopologyRfChange:
+    """
+        If any keyspace RF equals to number-of-cluster-nodes, where tablets are in use,
```
this docstring explains something that sound like a function.

why do we need a class for this ?

the name of it is quite vague, and doesn't state any responsibility this part has (or any actions)

yarongilor replied:
fixed by modifying:
```
class DataCenterTopologyRfControl:
    """
    This class manages and controls the replication factor (RF) of keyspaces in a ScyllaDB data center, when nodes are removed or re-added to the cluster.

    **Purpose**:
    - In scenarios where a keyspace has an RF equal to the total number of nodes in a data center, decommissioning a node is not supported where tablets are used.
    - This class provides functionality to temporarily decrease the RF of such keyspaces before a node decommissioning operation and revert them back to their original RF after a new node is added.

    **Usage**:
    1. **`decrease_keyspaces_rf`**: Identifies keyspaces with RF equal to the total number of nodes in the data center and decreases their RF by 1. This is necessary so decommissioning a node is allowed (with tablets).
    2. **`revert_to_original_keyspaces_rf`**: Reverts the RF of the keyspaces back to their original values after a new node is added to the data center.

    Attributes:
    - `target_node`: The node to decommission.
    - `datacenter`: The data center to which the target node belongs.
    - `decreased_rf_keyspaces`: A list of keyspaces whose RF has been decreased.
    - `original_nodes_number`: The original number of nodes in the data center (before decommission).
    """
```

yarongilor commented:
> > @fruch , @roydahan , @soyacz , can you please review the last 2 commits of this PR? it already encountered 2 new severe issues: [scylladb/scylladb#20282](https://github.com/scylladb/scylladb/issues/20282) , [scylladb/scylladb#20240](https://github.com/scylladb/scylladb/issues/20240).
> > A [test for decommission-related nemeses](https://argus.scylladb.com/test/e6972842-591d-472f-9e39-f196d3670053/runs?additionalRuns%5B%5D=4f0cedfd-0e8a-4bfd-9de7-115a32ff93da) passed ok, testing a single DC configuration. It had 3 nodes and RF=3 for stress keyspaces, so on each nemesis, the RF value was altered to 2, then back to 3 after a new node is added.
> > Testing a decommission in multi-DC seems to be blocked by the 2 above issues. Commits details:
> > 
> > * In case tablets are enabled, decommission is not permitted if a keyspace
> >   replicatin factor equals number-of-nodes.
> >   In this case the keyspace replication-factor should be decreased first.
> > * When replication-factor is increased after a new node was added,
> >   wait for tablets migration to complete.
> > 
> > cc: @bhalevy
> 
> @yarongilor just to clarify, those two issues doesn't happens without those changes ?

@fruch , right, this scenario has never be covered in SCT.

yarongilor commented:
> > > > @fruch , @roydahan , @soyacz , can you please review the last 2 commits of this PR? it already encountered 2 new severe issues: [scylladb/scylladb#20282](https://github.com/scylladb/scylladb/issues/20282) , [scylladb/scylladb#20240](https://github.com/scylladb/scylladb/issues/20240).
> > > > A [test for decommission-related nemeses](https://argus.scylladb.com/test/e6972842-591d-472f-9e39-f196d3670053/runs?additionalRuns%5B%5D=4f0cedfd-0e8a-4bfd-9de7-115a32ff93da) passed ok, testing a single DC configuration. It had 3 nodes and RF=3 for stress keyspaces, so on each nemesis, the RF value was altered to 2, then back to 3 after a new node is added.
> > > > Testing a decommission in multi-DC seems to be blocked by the 2 above issues. Commits details:
> > > > 
> > > > * In case tablets are enabled, decommission is not permitted if a keyspace
> > > >   replicatin factor equals number-of-nodes.
> > > >   In this case the keyspace replication-factor should be decreased first.
> > > > * When replication-factor is increased after a new node was added,
> > > >   wait for tablets migration to complete.
> > > > 
> > > > cc: @bhalevy
> > > 
> > > 
> > > @yarongilor just to clarify, those two issues doesn't happens without those changes ?
> > 
> > 
> > @fruch , right, this scenario has never been covered in SCT, it's new.
> 
> that complicates things a bit we shouldn't merge until those are fixed, to not waste peoples times on it.
> 
> on the other had this PR original was to attend to an issue also cause people to spend time figuring out.
> 
> seems like the crush is happening only ipv6, can you try on a different test case that doesn't use ipv6 ?
> 
> if both issue are ipv6 related, that it's only one case in tier1, and we can live with that, until those would be addressed.

[retested](https://argus.scylladb.com/test/e6972842-591d-472f-9e39-f196d3670053/runs?additionalRuns[]=5c8d1792-9a85-4b30-a482-cb6b84c09e90) without ipv6.

fruch commented on sdcm/utils/database_query_utils.py:
```diff
@@ -190,6 +190,11 @@ def get_table_clustering_order(ks_cf: str, ck_name: str, session) -> str:
     return clustering_order
 
 
+def is_system_keyspace(keyspace: str) -> bool:
```
how this change is related to this PR ?

I would expect it to be at the minimum on it's own commit

fruch commented on sdcm/utils/database_query_utils.py:
```diff
@@ -190,6 +190,11 @@ def get_table_clustering_order(ks_cf: str, ck_name: str, session) -> str:
 
 
+def is_system_keyspace(keyspace: str) -> bool:
+    system_keyspace_prefixes = ("system", "alternator_usertable", "audit")
```
`alternator_usertable` isn't really a system keyspace...

fruch commented:
> > > > > @fruch , @roydahan , @soyacz , can you please review the last 2 commits of this PR? it already encountered 2 new severe issues: [scylladb/scylladb#20282](https://github.com/scylladb/scylladb/issues/20282) , [scylladb/scylladb#20240](https://github.com/scylladb/scylladb/issues/20240).
> > > > > A [test for decommission-related nemeses](https://argus.scylladb.com/test/e6972842-591d-472f-9e39-f196d3670053/runs?additionalRuns%5B%5D=4f0cedfd-0e8a-4bfd-9de7-115a32ff93da) passed ok, testing a single DC configuration. It had 3 nodes and RF=3 for stress keyspaces, so on each nemesis, the RF value was altered to 2, then back to 3 after a new node is added.
> > > > > Testing a decommission in multi-DC seems to be blocked by the 2 above issues. Commits details:
> > > > > 
> > > > > * In case tablets are enabled, decommission is not permitted if a keyspace
> > > > >   replicatin factor equals number-of-nodes.
> > > > >   In this case the keyspace replication-factor should be decreased first.
> > > > > * When replication-factor is increased after a new node was added,
> > > > >   wait for tablets migration to complete.
> > > > > 
> > > > > cc: @bhalevy
> > > > 
> > > > 
> > > > @yarongilor just to clarify, those two issues doesn't happens without those changes ?
> > > 
> > > 
> > > @fruch , right, this scenario has never been covered in SCT, it's new.
> > 
> > 
> > that complicates things a bit we shouldn't merge until those are fixed, to not waste peoples times on it.
> > on the other had this PR original was to attend to an issue also cause people to spend time figuring out.
> > seems like the crush is happening only ipv6, can you try on a different test case that doesn't use ipv6 ?
> > if both issue are ipv6 related, that it's only one case in tier1, and we can live with that, until those would be addressed.
> 
> [retested](https://argus.scylladb.com/test/e6972842-591d-472f-9e39-f196d3670053/runs?additionalRuns%5B%5D=5c8d1792-9a85-4b30-a482-cb6b84c09e90) without ipv6 ==> [scylladb/scylladb#20282](https://github.com/scylladb/scylladb/issues/20282) is reproduced without ipv6.
> 
> @fruch , the issues are only reproduced with multi-dc. So it may still be reasonable to merge this PR that will solve all single-dc longevity decommission failures.

we have one multi-dc case in tier1, so I'll assign you to investigate it for a next runs after this merge.

yarongilor commented:
> @yarongilor is alter keyspace supported in 6.0,6.1,2024.2 ?

should be supported.

fruch commented:
> > @yarongilor is alter keyspace supported in 6.0,6.1,2024.2 ?
> 
> should be supported.

good, so go over the backports and fix the conflicts for this one.

yarongilor commented:
A decommission passed ok in master tier-1 [large-partitions-4k-pk](https://argus.scylladb.com/test/996501ad-c137-40ef-8dcc-a7d12f0ba3b6/runs?additionalRuns[]=c378699d-ae78-431f-b299-ba647bda7e0c)
